### PR TITLE
Fix compatibility with Doctrine DBAL 2.13

### LINKED ISSUE #549: Compatibilty with  Doctrine DBAL 2.13 broken
#### Sonata packages

<details><summary>show</summary>
<p>

```
$ composer show --latest 'sonata-project/*'
# Put the result here.
```
sonata-project/block-bundle          3.23.3 4.8.0  Symfony SonataBlockBundle
sonata-project/cache                 2.2.0  2.2.0  Cache library
sonata-project/classification-bundle 3.18.0 3.18.0 Symfony SonataClassificationBundle
sonata-project/datagrid-bundle       3.4.1  3.4.2  Symfony SonataDatagridBundle
sonata-project/doctrine-extensions   1.15.0 1.15.0 Doctrine2 behavioral extensions
sonata-project/exporter              2.9.0  2.9.0  Lightweight Exporter library
sonata-project/form-extensions       1.12.1 1.12.1 Symfony form extensions
sonata-project/media-bundle          3.36.0 3.36.0 Symfony SonataMediaBundle
sonata-project/twig-extensions       1.9.0  1.9.0  Sonata twig extensions

</p>
</details>

#### Symfony packages

<details><summary>show</summary>
<p>

```
$ composer show --latest 'symfony/*'
# Put the result here.
```
symfony/asset                      v4.4.27 v5.3.4  Manages URL generation and versioning of web assets such as CSS stylesheets, JavaScript files and image files
symfony/browser-kit                v4.4.27 v5.3.4  Simulates the behavior of a web browser, allowing you to make requests, click on links and submit forms programmati...
symfony/cache                      v4.4.33 v5.3.10 Provides an extended PSR-6, PSR-16 (and tags) implementation
symfony/cache-contracts            v2.4.0  v2.4.0  Generic abstractions related to caching
symfony/config                     v4.4.33 v5.3.10 Helps you find, load, combine, autofill and validate configuration values of any kind
symfony/console                    v4.4.33 v5.3.10 Eases the creation of beautiful and testable command line interfaces
symfony/css-selector               v4.4.27 v5.3.4  Converts CSS selectors to XPath expressions
symfony/debug                      v4.4.31 v4.4.31 Provides tools to ease debugging PHP code
symfony/debug-bundle               v4.4.27 v5.3.4  Provides a tight integration of the Symfony Debug component into the Symfony full-stack framework
symfony/dependency-injection       v4.4.33 v5.3.10 Allows you to standardize and centralize the way objects are constructed in your application
symfony/deprecation-contracts      v2.4.0  v2.4.0  A generic function and convention to trigger deprecation notices
symfony/doctrine-bridge            v4.4.31 v5.3.8  Provides integration for Doctrine with various Symfony components
symfony/dom-crawler                v4.4.30 v5.3.7  Eases DOM navigation for HTML and XML documents
symfony/dotenv                     v4.4.33 v5.3.10 Registers environment variables from a .env file
symfony/error-handler              v4.4.30 v5.3.7  Provides tools to manage errors and ease debugging PHP code
symfony/event-dispatcher           v4.4.30 v5.3.7  Provides tools that allow your application components to communicate with each other by dispatching events and list...
symfony/event-dispatcher-contracts v1.1.9  v2.4.0  Generic abstractions related to dispatching event
symfony/expression-language        v4.4.30 v5.3.7  Provides an engine that can compile and evaluate expressions
symfony/filesystem                 v4.4.27 v5.3.4  Provides basic utilities for the filesystem
symfony/finder                     v4.4.30 v5.3.7  Finds files and directories via an intuitive fluent interface
symfony/flex                       v1.17.2 v1.17.2 Composer plugin for Symfony
symfony/form                       v4.4.33 v5.3.10 Allows to easily create, process and reuse HTML forms
symfony/framework-bundle           v4.4.31 v5.3.10 Provides a tight integration between Symfony components and the Symfony full-stack framework
symfony/http-client                v4.4.33 v5.3.10 Provides powerful methods to fetch HTTP resources synchronously or asynchronously
symfony/http-client-contracts      v2.4.0  v2.4.0  Generic abstractions related to HTTP clients
symfony/http-foundation            v4.4.33 v5.3.10 Defines an object-oriented layer for the HTTP specification
symfony/http-kernel                v4.4.33 v5.3.10 Provides a structured process for converting a Request into a Response
symfony/inflector                  v4.4.27 v5.3.4  Converts words between their singular and plural forms (English only)
Package symfony/inflector is abandoned, you should avoid using it. Use use `EnglishInflector` from the String component instead instead.
symfony/intl                       v4.4.31 v5.3.8  Provides a PHP replacement layer for the C intl extension that includes additional data from the ICU library
symfony/mailer                     v4.4.27 v5.3.9  Helps sending emails
symfony/maker-bundle               v1.34.1 v1.34.1 Symfony Maker helps you create empty commands, controllers, form classes, tests and more so you can forget about wr...
symfony/mercure                    v0.4.1  v0.6.0  Symfony Mercure Component
symfony/mercure-bundle             v0.2.6  v0.3.3  Symfony MercureBundle
symfony/mime                       v4.4.31 v5.3.8  Allows manipulating MIME messages
symfony/monolog-bridge             v5.2.12 v5.3.7  Provides integration for Monolog with various Symfony components
symfony/monolog-bundle             v3.7.1  v3.7.1  Symfony MonologBundle
symfony/options-resolver           v4.4.30 v5.3.7  Provides an improved replacement for the array_replace PHP function
symfony/phpunit-bridge             v5.3.10 v5.3.10 Provides utilities for PHPUnit, especially user deprecation notices management
symfony/polyfill-apcu              v1.23.0 v1.23.0 Symfony polyfill backporting apcu_* functions to lower PHP versions
symfony/polyfill-ctype             v1.23.0 v1.23.0 Symfony polyfill for ctype functions
symfony/polyfill-iconv             v1.23.0 v1.23.0 Symfony polyfill for the Iconv extension
symfony/polyfill-intl-grapheme     v1.23.1 v1.23.1 Symfony polyfill for intl's grapheme_* functions
symfony/polyfill-intl-icu          v1.23.0 v1.23.0 Symfony polyfill for intl's ICU-related data and classes
symfony/polyfill-intl-idn          v1.23.0 v1.23.0 Symfony polyfill for intl's idn_to_ascii and idn_to_utf8 functions
symfony/polyfill-intl-normalizer   v1.23.0 v1.23.0 Symfony polyfill for intl's Normalizer class and related functions
symfony/polyfill-mbstring          v1.23.1 v1.23.1 Symfony polyfill for the Mbstring extension
symfony/polyfill-php56             v1.20.0 v1.20.0 Symfony polyfill backporting some PHP 5.6+ features to lower PHP versions
symfony/polyfill-php72             v1.23.0 v1.23.0 Symfony polyfill backporting some PHP 7.2+ features to lower PHP versions
symfony/polyfill-php73             v1.23.0 v1.23.0 Symfony polyfill backporting some PHP 7.3+ features to lower PHP versions
symfony/polyfill-php80             v1.23.1 v1.23.1 Symfony polyfill backporting some PHP 8.0+ features to lower PHP versions
symfony/polyfill-php81             v1.23.0 v1.23.0 Symfony polyfill backporting some PHP 8.1+ features to lower PHP versions
symfony/process                    v4.4.30 v5.3.7  Executes commands in sub-processes
symfony/property-access            v4.4.30 v5.3.8  Provides functions to read and write from/to an object or array using a simple string notation
symfony/property-info              v4.4.31 v5.3.8  Extracts information about PHP class' properties using metadata of popular sources
symfony/proxy-manager-bridge       v4.4.18 v5.3.4  Symfony ProxyManager Bridge
symfony/routing                    v4.4.30 v5.3.7  Maps an HTTP request to a set of configuration variables
symfony/security-bundle            v4.4.27 v5.3.8  Provides a tight integration of the Security component into the Symfony full-stack framework
symfony/security-core              v4.4.33 v5.3.10 Symfony Security Component - Core Library
symfony/security-csrf              v4.4.27 v5.3.4  Symfony Security Component - CSRF Library
symfony/security-guard             v4.4.27 v5.3.7  Symfony Security Component - Guard
symfony/security-http              v4.4.30 v5.3.10 Symfony Security Component - HTTP Integration
symfony/serializer                 v4.4.33 v5.3.10 Handles serializing and deserializing data structures, including object graphs, into array structures or other form...
symfony/service-contracts          v2.4.0  v2.4.0  Generic abstractions related to writing services
symfony/stopwatch                  v4.4.27 v5.3.4  Provides a way to profile code
symfony/string                     v5.3.10 v5.3.10 Provides an object-oriented API to strings and deals with bytes, UTF-8 code points and grapheme clusters in a unifi...
symfony/swiftmailer-bundle         v3.5.2  v3.5.2  Symfony SwiftmailerBundle
symfony/templating                 v4.4.30 v5.3.7  Provides all the tools needed to build any kind of template system
symfony/translation                v4.4.32 v5.3.10 Provides tools to internationalize your application
symfony/translation-contracts      v2.4.0  v2.4.0  Generic abstractions related to translation
symfony/twig-bridge                v4.4.27 v5.3.7  Provides integration for Twig with various Symfony components
symfony/twig-bundle                v4.4.30 v5.3.10 Provides a tight integration of Twig into the Symfony full-stack framework
symfony/validator                  v4.4.33 v5.3.10 Provides tools to validate values
symfony/var-dumper                 v4.4.33 v5.3.10 Provides mechanisms for walking through any arbitrary PHP variable
symfony/var-exporter               v4.4.31 v5.3.8  Allows exporting any serializable PHP data structure to plain PHP code
symfony/web-link                   v4.4.27 v5.3.4  Manages links between resources
symfony/web-profiler-bundle        v4.4.31 v5.3.8  Provides a development tool that gives detailed information about the execution of any request
symfony/webpack-encore-bundle      v1.12.0 v1.12.0 Integration with your Symfony app & Webpack Encore!
symfony/workflow                   v4.4.27 v5.3.4  Provides tools for managing a workflow or finite state machine
symfony/yaml                       v4.4.29 v5.3.6  Loads and dumps YAML files

</p>
</details>

#### PHP version

```
$ php -v
# Put the result here.
```
PHP 7.3.31-1+0~20210923.88+debian9~1.gbpac4058 (cli) (built: Sep 23 2021 22:05:13) ( NTS )
Copyright (c) 1997-2018 The PHP Group
Zend Engine v3.3.31, Copyright (c) 1998-2018 Zend Technologies
    with Zend OPcache v7.3.31-1+0~20210923.88+debian9~1.gbpac4058, Copyright (c) 1999-2018, by Zend Technologies
    with Xdebug v3.0.4, Copyright (c) 2002-2021, by Derick Rethans

## Steps to reproduce

Performing an iteration with "DoctrineDBALConnectionSourceIterator

## Actual results

```
In DoctrineDBALConnectionSourceIterator.php line 65:
                                                        
  [Error]                                               
  Call to a member function fetchAssociative() on bool  
```                                                

In version 3 $statement->execute($this->parameters) return a Result, but it was a boolean in previous versions 

https://github.com/doctrine/dbal/blob/2.13.x/lib/Doctrine/DBAL/Driver/Statement.php#L81-L95

### PR DESCRIPTION
**Pull request by jordisala1991**
## Subject

I am targeting this branch, because this is BC.

Closes #549 .

## Changelog

```markdown
### Fixed
- Fixed compatibility with Doctrine DBAL ^2.13.
```

### CODE REVIEW

**Comment by jordisala1991** [comment]
Can you confirm it works for you? @jerome-fix

File: src/Source/DoctrineDBALConnectionSourceIterator.php [hunk 1]
```diff
@@ -79,7 +79,14 @@ public function valid(): bool
     public function rewind(): void
     {
         $statement = $this->connection->prepare($this->query);
-        $this->result = $statement->execute($this->parameters);
+
+        if (method_exists($statement, 'executeQuery')) {
```
Comment (VincentLanglet) [inline src/Source/DoctrineDBALConnectionSourceIterator.php:84]: Can you add a comment about when dropping the check ?

Comment (jordisala1991) [inline src/Source/DoctrineDBALConnectionSourceIterator.php:84]:
```suggestion
        // TODO: Keep only the if part when dropping support for Doctrine DBAL < 3.1
        if (method_exists($statement, 'executeQuery')) {
```

Comment (jordisala1991) [reply src/Source/DoctrineDBALConnectionSourceIterator.php:84]: Done

**Review by VincentLanglet — APPROVED** [review]

**Review by core23 — APPROVED** [review]

**Comment by jerome-fix** [comment]
> Can you confirm it works for you? @jerome-fix

It works on my side.
Thanks a lot !

**Pull request merged**